File: sjf_scheduler.py
from .scheduler import Scheduler
from .process import Process
from typing import List
import copy

class SJFScheduler(Scheduler):
    """Shortest Job First Scheduler Implementation"""
    
    def schedule(self) -> None:
        # Create a copy of processes to avoid modifying original arrival times
        remaining_processes = [copy.deepcopy(p) for p in self.processes]
        ready_queue = []
        self.current_time = 0
        self.execution_sequence = []
        
        # Sort initially by arrival time
        remaining_processes.sort(key=lambda x: x.arrival_time)
        
        while remaining_processes or ready_queue:
            # Add all processes that have arrived to ready queue
            while remaining_processes and remaining_processes[0].arrival_time <= self.current_time:
                ready_queue.append(remaining_processes.pop(0))
            
            if not ready_queue:
                # No process available, jump to next arrival time
                self.current_time = remaining_processes[0].arrival_time
                continue
            
            # Sort ready queue by burst time (shortest job first)
            ready_queue.sort(key=lambda x: x.burst_time)
            
            # Get shortest job
            current_process = ready_queue.pop(0)
            
            # Find original process to update its times
            original_process = next(p for p in self.processes 
                                 if p.pid == current_process.pid)
            
            # Set start time if not already set
            if original_process.start_time is None:
                original_process.start_time = self.current_time
            
            # Execute process
            start = self.current_time
            self.current_time += current_process.burst_time
            
            # Update completion time
            original_process.completion_time = self.current_time
            
            # Add to execution sequence
            self.execution_sequence.append((original_process, start, self.current_time))

File: SJFScheduler.java
package com.cpuscheduler.core;

import com.cpuscheduler.model.Process;
import java.util.*;

/**
 * Shortest Job First (SJF) CPU scheduling algorithm (non-preemptive).
 * The process with the shortest burst time is selected for execution next.
 */
public class SJFScheduler extends AbstractScheduler {

    @Override
    public String getName() {
        return "Shortest Job First (SJF) - Non-preemptive";
    }

    @Override
    public List<String> execute() {
        // Reset state
        ganttChart.clear();
        completedProcesses.clear();
        currentTime = 0;
        contextSwitches = 0;

        // Make a copy of processes to avoid modifying the original list
        List<Process> remainingProcesses = new ArrayList<>(processes);
        
        // Sort processes by arrival time initially
        remainingProcesses.sort(Comparator.comparingInt(Process::getArrivalTime));

        while (!allProcessesCompleted()) {
            // Get all arrived processes that haven't completed yet
            List<Process> arrivedProcesses = new ArrayList<>();
            for (Process p : remainingProcesses) {
                if (p.getArrivalTime() <= currentTime && !p.isCompleted()) {
                    arrivedProcesses.add(p);
                }
            }

            if (arrivedProcesses.isEmpty()) {
                // No process is ready to execute, increment time
                currentTime++;
                continue;
            }

            // Find the process with the shortest burst time
            Process currentProcess = arrivedProcesses.stream()
                    .min(Comparator.comparingInt(Process::getBurstTime))
                    .orElse(null);

            if (currentProcess != null) {
                // Execute the process until completion (non-preemptive)
                int startTime = currentTime;
                int executionTime = currentProcess.getBurstTime();
                currentTime += executionTime;
                
                // Update process completion
                currentProcess.execute(executionTime, startTime);
                completedProcesses.add(currentProcess);
                
                // Update Gantt chart
                updateGanttChart(currentProcess, startTime, currentTime);
                
                // Count context switch (only if there are more processes to come)
                if (!allProcessesCompleted()) {
                    contextSwitches++;
                }
            } else {
                currentTime++;
            }
        }

        return ganttChart;
    }
}
